### ISSUE TITLE
Issue with .raw method and git cherry-pick

### ISSUE THREAD

**Issue by thecodechemist99** [steveukx/git-js#656]
Hi,
I found this solution in an old closed issue:

---
There isn't currently development on supporting `cherry-pick`, but you can use the library's `raw` method to still run your command at the moment:

```
const git = require('simple-git');
git().raw(['cherry-pick', '--strategy=recursive', '-X', 'theirs', sha], function (err, data) {
  err && console.error(err);
  err || console.log(data);
}); 
```

_Originally posted by @steveukx in https://github.com/steveukx/git-js/issues/270#issuecomment-390886146_

---

Unfortunately, I’m having issues with this approach (working in typescript). My compiler throws this error:
```
src/main.ts:206:7 - error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type 'string[]' is not assignable to parameter of type 'string'.

206       ['cherry-pick', commitHash],
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/simple-git/typings/simple-git.d.ts:470:4
    470    raw(a: string, b: string, callback?: types.SimpleGitTaskCallback<string>): Response<string>;
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The last overload is declared here.


Found 1 error.
```

This is my code (all the presumably necessary bits), changed it a little bit, but even the same code you suggested above throws the same error:
```
import fsSync from 'fs';
import simpleGit, { SimpleGit, SimpleGitOptions } from 'simple-git';

const gitOptions: SimpleGitOptions = {
  baseDir: `${process.cwd()}/docs`,
  binary: 'git',
  maxConcurrentProcesses: 6,
  config: []
};

const git = gitSetup(gitOptions);

function gitSetup (options: SimpleGitOptions): SimpleGit {
  if (!fsSync.existsSync('./docs/.git')) {
    fsSync.mkdirSync('./docs', { recursive: true });
  }
  return simpleGit(options);
}

// ...

async function publish (filePath: string): Promise<void> {
  // get commit hash
  const commitHash = (await git.log({ file: `./docs/${filePath}` })).latest.hash;

  // publish
  console.log('Publishing ...');
  const status = await git
    .checkout('publish-test')
    .pull()
    .raw(
      ['cherry-pick', commitHash],
      (err, data) => {
        if (err) {
          console.error(err);
        } else {
          console.log(data);
        }
      }
    )
    .push('origin', 'publish-test');
  console.log(status);
  console.log('Publishing completed.');
}
```

I even found this overload in the simple-git.d.ts, I wonder why this doesn’t work out:
```
raw(commands: string | string[] | types.TaskOptions, callback?: types.SimpleGitTaskCallback<string>): Response<string>;
```

Can you help with this issue?

**Comment by steveukx** [comment]
I am able to compile your code above with `typescript@4.1.2` and `simple-git@2.41.1` but needed to change the `commitHash` to be null-safe:

```
// get commit hash
const commitHash = (await git.log({ file: `./docs/${filePath}` })).latest?.hash;
```

Which TS/SG versions are you using?

**Comment by thecodechemist99** [comment]
I'm using typescript@4.3.5 and simple-git@2.41.1.

For me, the null-safe modification didn‘t solve the issue, unfortunately.

**Comment by thecodechemist99** [comment]
Sorry for the delay, I only work once a week on this projekt.
I tried to compile it with `typescript@4.1.2` and a completely reinstalled `simple-git@2.42.0`, as well as `simple-git@2.41.1` again. Didn’t compile, same error.

Might this in any way be related to my typescript options?
Here’s my tsconfig.json:
```
{
  "compilerOptions": {
      "target": "ES6",
      "rootDir": "src",
      "moduleResolution": "node",
      "allowSyntheticDefaultImports": true
  }
}
```

**Comment by thecodechemist99** [comment]
I’m still having issues with this.
I tried to compile just exactly the code I provided you above instead of my whole project. It still doesn’t work, I get the same error.

Here’s exactly what I did:
```
1. Create main.ts with the code above and your suggested change
2. npm init
3. npm install simple-git@2.41.1
4. npm install typescript@4.1.2
5. npm i --save-dev @types/node
6. Add tsconfig.json
7. tsc
```

tsconfig.json:
```
{
  "compilerOptions": {
      "target": "ESNext",
      "moduleResolution": "node",
      "allowSyntheticDefaultImports": true
  }
}
```

package.json:
```
{
  "name": "simple-git-debug",
  "version": "1.0.0",
  "description": "",
  "main": "main.js",
  "scripts": {
    "start": "tsc && node main.js"
  },
  "author": "",
  "license": "MIT",
  "dependencies": {
    "simple-git": "^2.41.1",
    "typescript": "^4.1.2"
  },
  "devDependencies": {
    "@types/node": "^16.4.13"
  }
}
```

What OS are you using? Not that this should make a difference, but you never know …
I’m on Windows 10 version 1909.

**Comment by steveukx** [comment]
Hi, thank you for all the additional detail on this.

There is a problem with the callback types - I'm using an interface with multiple call signatures rather than a union type of function types, so `Parameters<SimpleGitTaskCallback<string>>` currently always comes back as just `[Error]` rather than `[null, string] | [Error]`.

I will arrange a fix for this, but in the meantime you can unblock yourself by explicitly adding types to your code, eg:


```
const status = await git
    .checkout('publish-test')
    .pull()
    .raw(['cherry-pick', commitHash], (...[err, data]: [Error] | [null, string]) => {
        if (err) console.error(err);
        if (data) console.log(data);
    })
   .push('origin', 'publish-test');
```

**Comment by steveukx** [comment]
Hi, you should now be able to upgrade to [2.43.0](https://github.com/steveukx/git-js/releases/tag/v2.43.0) to use your existing code without the need to explicitly set types.

If you have any other problems, please do open a new issue. Thanks.
